File: datastructure&algorithm/ds_algo_java/src/daily/y_2022/m_11/w_03/d_2022_11_15/bj_1520_2/Main.java
package daily.y_2022.m_11.w_03.d_2022_11_15.bj_1520_2;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static int[] dr = {-1,0,1,0};
    static int[] dc = {0,-1,0,1};

    public static void main(String[] args) throws IOException {
        StringTokenizer st = new StringTokenizer(br.readLine());
        int m = Integer.parseInt(st.nextToken());
        int n = Integer.parseInt(st.nextToken());

        int[][] map = new int[m+1][n+1];
        int[][] dp = new int[m+1][n+1];

        for(int i=1;i<=m;i++){
            for(int j=1;j<=n;j++){
                map[i][j] = 0;
                dp[i][j]=-1;
            }
        }
        for(int i=1;i<=m;i++){
            st = new StringTokenizer(br.readLine());
            for(int j=1;j<=n;j++){
                map[i][j] = Integer.parseInt(st.nextToken());

            }
        }

        System.out.println(dfs(1,1,map,dp,m,n));
    }

    static int dfs(int sr, int sc, int[][] map, int[][] dp,int m, int n){
        // printArr(dp);
        if(sr==m && sc ==n) return 1;
        if(dp[sr][sc]!=-1) return dp[sr][sc];

        dp[sr][sc] = 0;
        for(int i=0;i<4;i++){
            int nr = sr+dr[i];
            int nc = sc+dc[i];


            if(nr<1||nr>m||nc<1||nc>n) continue;

            if(map[nr][nc]>=map[sr][sc]) continue;


            dp[sr][sc]+=dfs(nr,nc,map,dp,m,n);
        }

        return dp[sr][sc];


    }
    static void printArr(int[][] map){
        System.out.println();
        for(int i=0;i<map.length;i++){
            for(int j=0;j<map[0].length;j++){
                System.out.printf("%3d ",map[i][j]);
            }
            System.out.println();
        }
    }
}

File: datastructure&algorithm/ds_algo_java/src/daily/y_2022/m_11/w_03/d_2022_11_15/bj_1520_2/Main.kt
package daily.y_2022.m_11.w_03.d_2022_11_15.bj_1520_2
import java.util.*


private val dr = arrayOf(-1,0,1,0)
private val dc = arrayOf(0,-1,0,1)

private var m =0
private var n = 0

private lateinit var map:Array<Array<Int>>
private lateinit var dp:Array<Array<Int>>

var br = System.`in`.bufferedReader()
fun main(){
    var st = StringTokenizer(algorithms.exhaustive_search.bj_1520_2.br.readLine())
    m = st.nextToken().toInt()
    n = st.nextToken().toInt()

    //맵
    map = Array(m+1){Array(n+1){0} }
    //거리
    dp = Array(m+1){Array(n+1){-1} }


    //맵 초기화
    repeat(m){i->
        st = StringTokenizer(algorithms.exhaustive_search.bj_1520_2.br.readLine())
        repeat(n){j->
            map[i+1][j+1] = st.nextToken().toInt()
        }
    }
    println(algorithms.exhaustive_search.bj_1520_2.dfs(1, 1))

}

fun dfs(sr:Int, sc:Int):Int{
    if(sr==m && sc==n) return 1
    if(dp[sr][sc]!=-1) return dp[sr][sc]

    dp[sr][sc]=0
    for(i in 0..3){
        val nr = sr+dr[i]
        val nc = sc+dc[i]

        if(nr<1||nr>m||nc<1||nc<n) continue
        if(map[sr][sr]<=map[nr][nc]) continue

        dp[sr][sc]+= algorithms.exhaustive_search.bj_1520_2.dfs(nr, nc)
    }
    return dp[sr][sc];
}